File: 01-programming/python/concepts/code/01-basics/syntax-variables.py
# -o-o-o-o- Syntax (Variables) -o-o-o-o-

# This file goes over variables in Python simply.
#
#  # # # # KEY # # # #
#
# variable   - a named storage location that holds a value which can change.
#              Same concept as C and C#, but with key differences.
#
# No type declaration - Python is dynamically typed. You do NOT write the type
#              when declaring a variable. Python figures it out automatically.
#              C:  int score = 0;
#              C#: int score = 0;
#              Py: score = 0
#
# =          - the assignment operator. Stores a value into a variable.
#              Same as in C and C#.
#
# type()     - a built-in function that returns the type of a variable.
#              eg. type(score) returns <class 'int'>
#              Useful for checking what type Python assigned automatically.
#
# Dynamic typing - a variable's type can change when reassigned to a different
#              type. eg. x = 10 then x = "hello" is valid in Python.
#              This is NOT allowed in C or C#.
#
# Naming convention - Python uses snake_case for variable names.
#              eg. my_variable, player_name, high_score
#              C# uses camelCase. C uses either.
#
# Scope      - a variable declared inside a function only exists there.
#              Same concept as C and C#.
#
# print()    - prints text to the console with a newline at the end.
# f"..."     - f-string. Embeds variables directly in a string.
#

# Integer variable (no type needed)
score = 0
print(f"Initial score: {score}")

score = 50          # reassigning
print(f"Updated score: {score}")

score = score + 10  # using the variable in an expression
print(f"Score after +10: {score}")

# Float variable
temperature = 36.6
print(f"\nTemperature: {temperature}")

temperature = temperature - 1.5
print(f"Temperature after -1.5: {temperature}")

# Boolean variable (True or False — capital first letter in Python)
is_running = True
print(f"\nProgram running: {is_running}")

is_running = False
print(f"Program running: {is_running}")

# String variable
player_name = "Hero"
print(f"\nPlayer: {player_name}")

player_name = "SuperHero"
print(f"Player: {player_name}")

# Checking types with type()
print(f"\ntype(score):       {type(score)}")
print(f"type(temperature): {type(temperature)}")
print(f"type(is_running):  {type(is_running)}")
print(f"type(player_name): {type(player_name)}")

# Dynamic typing — a variable can change type (unique to Python)
x = 10
print(f"\nx is: {x} — type: {type(x)}")
x = "now a string"
print(f"x is: {x} — type: {type(x)}")

# Multiple assignment in one line
a, b, c = 1, 2, 3
print(f"\na={a}, b={b}, c={c}")

File: 01-programming/csharp/concepts/code/01-basics/syntax-variables.cs
// -o-o-o-o- Syntax (Variables) -o-o-o-o-

/* This file goes over variables in C# simply.
 *
 *  # # # # KEY # # # #
 *
 * variable - a named storage location that holds a value which can change
 *            during the program. Same concept as in C.
 *
 * type     - every variable has a type that defines what kind of data it stores.
 *            C# is strongly typed — you cannot store a string in an int, etc.
 *            Same concept as C.
 *
 * =        - the assignment operator. Stores a value into a variable.
 *
 * Naming rules (C# and C share most of these):
 *   - Must start with a letter or underscore (_), not a number.
 *   - Cannot use reserved keywords (eg. int, class, return).
 *   - Case sensitive: myVar and myvar are different variables.
 *   - C# convention: use camelCase for local variables. eg. myVariable.
 *
 * Scope    - where a variable can be accessed. A variable declared inside {}
 *            only exists within those braces. Same concept as C.
 *
 * Console.WriteLine() - prints text to the console with a newline at the end.
 *
 * Differences from C:
 *   - C# has 'string' as a built-in type; C uses char arrays.
 *   - C# has 'bool' with true/false; C uses int (1/0).
 *   - C# variables cannot be used before being assigned (local scope).
 *   - C# supports 'var' for implicit typing (see syntax-init.cs).
 *
 * */

using System;

class Program {
    static void Main() {

        // Integer variable
        int score = 0;
        Console.WriteLine($"Initial score: {score}");

        score = 50; // reassigning
        Console.WriteLine($"Updated score: {score}");

        score = score + 10; // using the variable in an expression
        Console.WriteLine($"Score after +10: {score}");

        // Double variable
        double temperature = 36.6;
        Console.WriteLine($"\nTemperature: {temperature}");

        temperature = temperature - 1.5;
        Console.WriteLine($"Temperature after -1.5: {temperature}");

        // Boolean variable (true or false)
        bool isRunning = true;
        Console.WriteLine($"\nProgram running: {isRunning}");

        isRunning = false;
        Console.WriteLine($"Program running: {isRunning}");

        // String variable
        string playerName = "Hero";
        Console.WriteLine($"\nPlayer: {playerName}");

        playerName = "SuperHero"; // strings can be reassigned
        Console.WriteLine($"Player: {playerName}");

        // Char variable (single character, uses single quotes)
        char grade = 'A';
        Console.WriteLine($"\nGrade: {grade}");

        // Demonstrating scope — this variable only exists inside these braces
        {
            int localVar = 999;
            Console.WriteLine($"\nInside scope: localVar = {localVar}");
        }
        // Console.WriteLine(localVar); // ERROR: localVar does not exist out here

    }
}
